developer_tools/malloc_dbg.c

```diff
@@ -0,0 +1,252 @@
+	const char* fileName; ///< File in which malloc or calloc has been called
+	const char* functionName; ///< Function in which malloc or calloc has been called
+	size_t bytes; ///< Number of bytes allocated
+	int line; ///< Line number (in file) corresponding to the malloc / calloc call
+	struct MEMORY_INFORMATION* next; ///< Next element in the list
+	struct MEMORY_INFORMATION* previous; ///< Previous element in the list
+} mem_info;
+
+/** We use a global variable for the list so we can use it at any time.
+ * */
+mem_info* memoryInformation = NULL;
+
+/** Another global variable. This one is used to know if we already call the atexit function.
+ * */
+int atexitCalled = 0;
+
+
+/**
+ * @brief addMemInfo function
+ * @details This function creates a new element and add it on top of the list
+ * @returns mem_info pointer if it succeeds, NULL otherwise
+ */
+mem_info* addMemInfo(mem_info* memoryInfo, void* ptrToReturn, size_t bytes, int line, const char* filename, const char* functionName)
+{
+	mem_info* newMemInfo = (mem_info*)malloc(sizeof(mem_info));
+	if (!newMemInfo)
+		return NULL;
+
+	newMemInfo->ptr = ptrToReturn;
+	newMemInfo->bytes = bytes;
+	newMemInfo->line = line;
+	newMemInfo->fileName = filename;
+	newMemInfo->functionName = functionName;
+	newMemInfo->next = memoryInfo;
+	newMemInfo->previous = NULL;
+	if (memoryInformation)
+		memoryInformation->previous = newMemInfo;
+
+	return newMemInfo;
+}
+
+/**
+ * @brief inList function
+ * @details This function is used to know if an allocation in a specific file at a specific line already exists in the list.
+ * @returns Position of the element in the list if the element is found, -1 otherwise.
+ */
+int inList(const char* filename, int line)
+{
+	mem_info* tmp = memoryInformation;
+	int counter = 0;
+	int len = strlen(filename);
+
+	while (tmp)
+	{
+		if (len == strlen(tmp->fileName))
+		{
+			if (!memcmp(filename, tmp->fileName, len) && tmp->line == line)
+				return counter;
+		}
+		tmp = tmp->next;
+		counter++;
+	}
+	return -1;
+}
+
+/**
+ * @brief editInfo function
+ * @details This function is used to edit the number of bytes allocated at a specific line.
+ * @returns Nothing.
+ */
+void editInfo(int elemPos, size_t bytes)
+{
+	int counter = 0;
+	mem_info* tmp = memoryInformation;
+
+	while (counter != elemPos)
+	{
+		tmp = tmp->next;
+		counter++;
+	}
+	tmp->bytes += bytes;
+}
+
+/**
+ * @brief malloc_dbg function
+ * @details This function calls malloc and allocates the number of bytes passed in the parameters.
+ * If the allocation succeeds then it add the pointer returned by malloc in the mem_info list.
+ * @returns Pointer returned by malloc if it's valid, NULL otherwhise.
+ */
+void* malloc_dbg(size_t bytes, int line, const char* filename, const char* functionName)
+{
+	void* ptrToReturn = malloc(bytes);
+	int pos = 0;
+	if (!ptrToReturn)
+		return NULL;
+
+	/* We must check atexitCalled value to know if we already called the function */

```

Comment:
Inside functions, we use normal comments (`//`).
Please check other parts of the code, too.